api/chat.go

```diff
@@ -45,61 +45,124 @@ func configGinChatRouter(router *gin.RouterGroup) {
-			}
+			log.Warn(err)
 			return
 		}
-		if broadcast, err := json.Marshal(chatForDb); err == nil {
-			broadcastStream(tumLiveContext.Stream.ID, broadcast)
+		if req.Type == "message" {
+			handleMessage(tumLiveContext, s, msg)
+		} else if req.Type == "like" {
+			handleLike(tumLiveContext, msg)
 		}
 	})
 }
 
+func handleLike(ctx tools.TUMLiveContext, msg []byte) {
+	var req likeReq
+	err := json.Unmarshal(msg, &req)
+	if err != nil {
+		return
+	}
+	err = dao.ToggleLike(ctx.User.ID, req.Id)
+	if err != nil {
+		log.WithError(err).Error("error liking/unliking message")
+		return
+	}
+	numLikes, err := dao.GetNumLikes(req.Id)
+	if err != nil {
+		log.WithError(err).Error("error getting num of chat likes")
+		return
+	}
+	broadcast := gin.H{
+		"likes": req.Id,
+		"num":   numLikes,
+	}
+	broadcastBytes, err := json.Marshal(broadcast)
+	if err != nil {
+		log.WithError(err).Warn("Can't marshal like message")
+		return
+	}
+	broadcastStream(ctx.Stream.ID, broadcastBytes)
+}
+
+func handleMessage(ctx tools.TUMLiveContext, session *melody.Session, msg []byte) {
+	var chat chatReq
+	if err := json.Unmarshal(msg, &chat); err != nil {
+		log.WithError(err).Error("error unmarshaling chat message")
+		return
+	}
+	if !ctx.Course.ChatEnabled {
+		return
+	}
+	uname := ctx.User.Name
+	if chat.Anonymous && ctx.Course.AnonymousChatEnabled {
+		uname = "Anonymous"
+	}
+	replyTo := sql.NullInt64{}
+	if chat.ReplyTo == 0 {
+		replyTo.Int64 = 0
+		replyTo.Valid = false
+	} else {
+		replyTo.Int64 = chat.ReplyTo
+		replyTo.Valid = true
+	}
+	chatForDb := model.Chat{
+		UserID:   strconv.Itoa(int(ctx.User.ID)),
+		UserName: uname,
+		Message:  chat.Msg,
+		StreamID: ctx.Stream.ID,
+		Admin:    ctx.User.ID == ctx.Course.UserID,
+		ReplyTo:  replyTo,
+	}
+	err := dao.AddMessage(&chatForDb)
+	if err != nil {
+		if errors.Is(err, model.ErrCooledDown) {
+			sendServerMessage("You are sending messages too fast. Please wait a bit.", TypeServerErr, session)
+		}
+		return
+	}
+	if broadcast, err := json.Marshal(chatForDb); err == nil {
+		broadcastStream(ctx.Stream.ID, broadcast)
+	}
+}
+
 func getMessages(c *gin.Context) {
 	foundContext, exists := c.Get("TUMLiveContext")
 	if !exists {
 		c.AbortWithStatus(http.StatusNotFound)
 		return
 	}
 	tumLiveContext := foundContext.(tools.TUMLiveContext)
-	c.JSON(http.StatusOK, tumLiveContext.Stream.Chats)
+	var uid uint = 0 // 0 = not logged in. -> doesn't match a user
+	if tumLiveContext.User != nil {
+		uid = tumLiveContext.User.ID
+	}
+	chats, err := dao.GetChats(uid, tumLiveContext.Stream.ID)
+	if err != nil {
+		c.AbortWithStatus(http.StatusInternalServerError)
+		return
+	}
+	c.JSON(http.StatusOK, chats)
 }
 
-type ChatReq struct {
+type wsReq struct {
+	Type string `json:"type"`
+}
+
+type chatReq struct {
+	wsReq
 	Msg       string `json:"msg"`
 	Anonymous bool   `json:"anonymous"`
 	ReplyTo   int64  `json:"replyTo"`
 }
+
+type likeReq struct {
+	wsReq
+	Id uint `json:"id"`
+}
+
 type ChatRep struct {
```

Comment:
Not related to this PR, but it would be good, if `ChatRep` had a comment and or a more verbose name.

Reply:
It's not used anymore, I removed it.